### DISCUSSION TITLE
NGINX Install

### DISCUSSION THREAD

**Discussion by SeanTurner71** [rdudhagra/Sand-Table#19] [Troubleshooting] [answered]
Running into several issues programming the Pi.  The main show stopper is installing NGINX.  After running "sudo apt-get install nginx" I get nothing but 404 errors.  Any insights?

**Comment by rdudhagra** [comment, marked as answer]
Hmm, few things to debug/paste the output of here...
1. I assume you followed the tutorial in the build guide...including modifying the config files exactly/etc
2. What version of Raspbian are you running? Buster should be fine
3. Can you try the following:
```bash
sudo systemctl restart nginx
sudo systemctl status nginx # paste the output of this here
journalctl -xe # paste the output of this here
```
This will tell us if nginx is starting correctly.
4. Can you confirm the ports set in `.env` in the website folder (should be port 3000)?
5. Is your server running? Run `ps aux | grep node` and see if anything pops up.

Will need more info to debug further, but hopefully this is a good start...let me know if you find the solution though!

**Reply by SeanTurner71** [reply]
I have a few more things I want to try but here are a few pics.  First one is what looks like errors during install.  second one is the original 404 errors when trying to nginx part,  Third is the output to the commands you asked about.  I have tried Pi OS with desktop and recommended software, Pi OS with desktop and lite.  I'll update if I find anything new.
![install](https://user-images.githubusercontent.com/86273998/123498571-c648c780-d5fe-11eb-98dc-7ab50028d3de.jpg)
![404 errors](https://user-images.githubusercontent.com/86273998/123498581-d8c30100-d5fe-11eb-8d26-8eb0d968c6e0.jpg)
![errors](https://user-images.githubusercontent.com/86273998/123498586-dc568800-d5fe-11eb-900f-5567a336f90f.jpg)

**Reply by rdudhagra** [reply]
Oh, didn’t understand your question at first.
Can you try running `sudo apt-get update` and reinstalling nginx (if you haven’t already)? Also are you on your home WiFi or on a school/company network that maybe has a more restrictive firewall?

**Reply by SeanTurner71** [reply]
That did it.  I think my Pihole setup might have been causing an issue.  The website is not working but I have a nginx splashscreen and I see errors in terminal involving ports so I can work on that.  Thanks again.  I'll let you know if I run into any other issues.

**Reply by rdudhagra** [reply]
Sounds good, no problem! Yea feel free to post more logs/screenshots/etc... if you run into anymore issues

**Comment by SeanTurner71** [comment]
OK.  Made it a bit further.  It will load a file and print without issue.  Can use the website no prob.  The problem is that around 50% of the tracks I try to load seem to crash the server.  If I just use the ip page I get a 503 error.  If I go to IP:3000 I get a network error.  I read on another post where that could be due to corrupt tracks.  Here are some "ps aux | grep node" shots before and after crash
![before crash](https://user-images.githubusercontent.com/86273998/124523017-39062f80-ddc3-11eb-9277-98905a87d42b.jpg)
![after crash](https://user-images.githubusercontent.com/86273998/124523026-3e637a00-ddc3-11eb-9c4d-c6bfe19e213e.jpg)

**Reply by rdudhagra** [reply]
Hmm, interesting. What's the file size of the tracks that fail to upload? Can you also send the output of `sudo systemctl status sandtable` right after you get a 503/network error? 

Also, would you be able to attach one of the tracks that fails to upload? I can try on my end and see if I get the same error.

**Reply by SeanTurner71** [reply]
![Screenshot 2021-07-05 194136](https://user-images.githubusercontent.com/86273998/124524495-42929600-ddc9-11eb-9d02-9bac88412e9b.png)
[inwards_sunburst.zip](https://github.com/rdudhagra/Sand-Table/files/6766499/inwards_sunburst.zip)

**Reply by rdudhagra** [reply]
Ah yea seems like you're running out of memory! I've never been able to process `inwards_sunburst.thr` on my table either...what I ended up doing was running the server on my Mac/PC, letting it process the file for however long it takes (it took a long time) and then just copy the resulting files onto the Pi's webserver.

**Reply by rdudhagra** [reply]
Another thing you could do is increase the SWAP memory size on the Pi, which should help alleviate some of these issues. See https://pimylifeup.com/raspberry-pi-swap-file/ for how to do this. However, I'd really not recommend this approach, since it will be **very slow** and degrade your Pi's SD card.

**Reply by SeanTurner71** [reply]
Thanks.  Does that happen with alot of tracks?  A few more questions while I have you here.  For doing our own tracks on https://sandify.org/ is there any thing I should avoid that would make a track fail?  You mentioned running the server on your pc to generate the files, is that something easy or does it have to be in a linux vm or something?  Thanks again. 
Edit.  Would running this on  a higher memory pi4 make a diff?

**Reply by rdudhagra** [reply]
1. In my experience, only `inwards_sunburst` has caused me issues, but I haven't made anything on sandify
2. Just avoid really long tracks. Remember that at the end of the day, the server is converting each line of the `.thr` file into many many gcode commands, and so for a long file the Pi will have to do an insane number of trig calculations. I'm sure there are many optimizations I could make to improve speed/memory efficiency, but I probably won't get around to making them for a while. That being said, feel free to poke around and make a pr if that's something you're interested in!
3. It's very easy. Just install NodeJS for your platform (very easy to google), download the repo somewhere, and `npm install` and `npm start` to run. Let me know if you run into problems with this though...there might be some quirks with `npm install` that I've forgotten about.
4. Probably? If you think the investment of a 4/8 GB Pi 4 is worth it, then yea I think it could make a difference. Although it'll still take forever to process, because of the CPU limitations.

**Reply by SeanTurner71** [reply]
It turns out I had a pi4 4gig laying around.  It ended out giving me a timeout error of some sort.  So the memory issue maybe was fixed but then another one pops up.   I guess I'll just stay away from that track for the near future.  Thanks again for helping me out.  Hopefully it all goes smoothy from here on out.
edit: Well on reboot it might have done it.  It seems to be printing something
![inwards](https://user-images.githubusercontent.com/86273998/124529559-d881ed00-ddd8-11eb-8ae4-8a62eea202be.jpg)

**Reply by rdudhagra** [reply]
Aha, ok yea the timeout error is what I've seen on my setup. Although, I find it weird that you've run into memory issues and I haven't (I'm using a Pi 3B+). 

Based on the screenshot you posted the webserver probably stopped processing half-way through. From the looks of it the `gcode` has been created, but not the preview images. So if you don't really care about those this should be fine.

No worries, glad I could help. Good luck on your table—would love to see some pictures when you're done!

**Reply by SeanTurner71** [reply]
Yea I have a Pi 3B+ also so that's weird.  I just moved the flash card to the Pi4.  No other changes.  The timeout was the 1800 seconds in nginx proxy settings I think.

**Reply by rdudhagra** [reply]
Yea so the way I currently have it the server will only respond once ALL the processing is done. So the NGINX proxy will definitely time out before some of these big `.thr` files finish. You could try making this timeout something much larger like `86400` (24 hrs) and see if that helps. I'm not sure if closing the website after the file has been uploaded to the server (no easy way to see when this happens, but try the network tab in your browser's *"inspect element"* panel) will cause the server-side processing to terminate—if it doesn't then you could just put a track to process overnight/etc.
